### PR TITLE
Implement Rust-based Database Access

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description

* We want to deprecate the AnkiDroid backend codebase (`libAnki`) and instead use Anki Desktop's code (`rslib`)
  * This allows us to quickly follow the changes made to Anki Desktop, and ensures feature parity 
* Usage of the Rust Backend requires an open collection
* An open collection locks the database
* Therefore the first step in using `rslib` is to transfer the database access to flow through `rslib`

## Approach
* We load a `.so` created in: https://github.com/david-allison-1/Anki-Android-Backend
* Add a Java bridge to the Rust (`anki-android-backend`) https://github.com/david-allison-1/Anki-Android-Backend/blob/rust-wip/docs/OVERVIEW.md
* Add an abstraction layer over the Rust/Java and call into it

## Thoughts
* Maybe we want performance tests on the database now - maybe later
  * Seems roughly equivalent/slightly faster than requery
  * Probably: Partially because we don't have a JNI boundary hit when we perform `get(String|Int etc...)()` on a cursor - logic is handled in the Java.

## Known issues 

* Harder to mock time-based dependencies (`Collection.crt`)
* Debugging over the Java/Rust boundary appears to be extremely difficult
* `openCollection` will open the database and create it with schema and indices. This will not happen in the Java side, so the database setup is slightly different.
* ⚠️ Potential issue with database upgrades to schema 11 if the user is using a really old AnkiDroid - unsure if we should ask the user to use an earlier version of AnkiDroid, or Anki Desktop for this.
* Nested database cursors are not supported: https://github.com/david-allison-1/Anki-Android-Backend/issues/23
* Low memory devices do not use limit/offset for database queries: https://github.com/david-allison-1/Anki-Android-Backend/issues/14
* Missing documentation/sources for the main methods: https://github.com/david-allison-1/Anki-Android-Backend/issues/15
* Hardcoded page size streaming data from the Rust of 1000 items per page
* Rust does not stream data from the database (but has no Heap size limit like a Java app does)
* This increases the APK size (~10MB) due to the size of the `.so`, improving this currently is beyond my skillset. 
  * We'll later be able to remove requery
  * We can consolidate some of our assets
  * Code will be removed from `libAnki`
  * Far in the future (pie in the sky), we might be able to use libraries linked to the Android framework

## How Has This Been Tested?

* Physical Device, API 21 emulator and API 25 emulator

## Learning

* https://github.com/mozilla/rust-android-gradle
* How the framework database classes work internally - needed to re-implement all of them
* SQLite conversion rules/data types 
* https://doc.rust-lang.org/cargo/index.html
* Allowing native libraries to run under Robolectric
   * Cross-compile, then `Runtime.load()` 
   * Cross-compilation of Rust for Win32/Apple/Linux
   * And how Apple cross-compilation breaks the TOS for the Apple SDK if not run on an Apple machine
* Gradle internals
* Environment setup via doctor scripts (TODO)
* JNI in general
* JNI with Rust:
  * https://docs.rs/jni/0.18.0/jni/ 
* Protobufs, and writing a custom protobuf compiler
   * https://developers.google.com/protocol-buffers/docs/overview
   * https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.descriptor
   * https://expobrain.net/2015/09/13/create-a-plugin-for-google-protocol-buffer/
* GitHub Actions usage
   * bash does not login by default 
* Uploading to Maven (Sonatype OSS)
  * Attempts to use JitPack
  * plugins.gradle.org seems much more modern, but didn't seem to have any mechanism to manually rotate keys. Support was slow/non-responsive to a query
* https://github.com/david-allison-1/Anki-Android-Backend/blob/rust-wip/docs/OVERVIEW.md
   * plantUML 

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

### CODE REVIEW

**Comment by dae** [comment]
By "Debugging over the Java/Rust boundary appears to be extremely difficult" I presume you mean the inability to step the Java debugger into Rust code? For issues that depend on internal transient state such as your DB layer, it would probably be helpful for developers to have an easy way to run AnkiDroid with a locally compiled version of the backend, which they can augment with print statements. But I suspect once the initial infrastructure is set up, the majority of issues that will come up will either solely depend on the input to a backend method, or the input combined with the currently open collection. It's not much different to debugging code that calls a web API - if you feed it good data and you get back bad results, you know the issue is not in your own code, and you can then take that input data and debug it in the other project, or report the issue upstream.

**Comment by david-allison-1** [comment]
TL;DR: Agreed with the final point. I doubt there's going to be much debugging required of the library (famous last words).

Println-based debugging may be necessary, but I feel it's YAGNI right now to get a working setup.

----

> By "Debugging over the Java/Rust boundary appears to be extremely difficult" I presume you mean the inability to step the Java debugger into Rust code? 

Yep, there's been successful work in getting LLDB working, but it's far from the mixed-mode debugging experience from other toolchains/mixed languages: https://github.com/mozilla/rust-android-gradle/issues/22. 

From vague readings about the quality of the debuggers, I'd only consider this as a last resort, println debugging is sadly likely going to be the way to go.

> For issues that depend on internal transient state such as your DB layer, it would probably be helpful for developers to have an easy way to run AnkiDroid with a locally compiled version of the backend, which they can augment with print statements. 

We're already printing the database queries being sent to the database to the debug logs. 

Instructions for a locally compiled backend are already in the PR (I've been doing most of my testing with a locally compiled backend); they could do with a little more polish to save a ~10LOC change while testing under Robolectric.

Haven't implemented printing to logcat from Rust yet, which would be probably be the best form of debugging that we get for a fair while.

> But I suspect once the initial infrastructure is set up, the majority of issues that will come up will either solely depend on the input to a backend method, or the input combined with the currently open collection. It's not much different to debugging code that calls a web API - if you feed it good data and you get back bad results, you know the issue is not in your own code, and you can then take that input data and debug it in the other project, or report the issue upstream.

Agreed - I'd be surprised if we find issues with the upstream code as it's battle-tested.

**Comment by mikehardy** [comment]
I actually prefer print debugging I find it much faster in combo with a test infrastructure than debugger fiddling personally :-), as long as it's possible it's fine by me.

**Comment by mikehardy** [comment]
The list of knowledge is quite humorous to me - you had to learn All The Things to do this thing. Even android dependency publishing, what a PITA!

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 1]
```diff
@@ -187,12 +193,13 @@ public String getString(Resources res) {
     private static final int UNDO_SIZE_MAX = 20;
 
     @VisibleForTesting
-    public Collection(Context context, DB db, String path, boolean server, boolean log, @NonNull Time time) {
+    public Collection(Context context, DB db, String path, boolean server, boolean log, @NonNull Time time, BackendFactory backendFactory) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: `@Nullable BackendFactory` - if rust loading fails this will be null, and it changes to DB which is marked Nullable

File: AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java [hunk 2]
```diff
@@ -65,8 +73,32 @@ public static Collection Collection(Context context, String path, boolean server
...
+        }
+
+        if (attemptBuildingWithRustBackend() && backendFactory != null) {
+            // Note: This will partially create the database
+            Timber.i("backend: open collection %s", path);
+            db = new DB(path, backendFactory);
+        } else {
+            Timber.i("backend: skipping open collection");
+            db = new DB(path);
+        }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java]: This feels awkward to me 🤔 
I'm thinking -

- rename `attemptBuildingWithBackend` to `useBackend`, change the static to `useBackend` to match - harmonize the state variable name and all API names
- the try/catch and if/else seem to be sharing a lot of condionality...the catch basically does logging then sets a flag that the whole if/else depends on - would it make sense to move the if part of the if/else into the success part of the try, then the else part into the catch? that seems more clear to me

File: AnkiDroid/build.gradle [hunk 3]
```diff
@@ -241,6 +241,18 @@ dependencies {
     implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
     implementation 'androidx.webkit:webkit:1.4.0'
 
+    // == Rust conversion (from Anki-Android-Backend on GitHub) ==
+    implementation 'com.google.protobuf:protobuf-java:3.13.0' // TODO: Confirm whether this is still needed
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:245]: seems best to confirm and remove comment or remove dependency before merging?

File: AnkiDroid/build.gradle [hunk 4]
```diff
@@ -241,6 +241,18 @@ dependencies {
...
+    // use a variable as we want both testing and implementation on the same version
+    String backendVersion = "0.0.5"
+    // build with ./gradlew rsdroid:assembleRelease
+    // In my experience, using `files()` currently requires a reindex operation, which is slow.
+    // implementation files("C:\\GitHub\\Rust-Test\\rsdroid\\build\\outputs\\aar\\rsdroid-release.aar")
+    implementation "io.github.david-allison-1:anki-android-backend:$backendVersion"
+    // build with ./gradlew rsdroid-testing:assembleRelease
+    // add a one-time call to Runtime.getRuntime().load(path); to test locally
+    // TODO: Add method in RustBackendLoader to load native library on disk
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:254]: Definitely want the ability to load it for robolectric testing locally without any non-bug-related code change

Correct if wrong but I'm imaging the flow to be

- get Anki-Android-Backend clean+up-to-date and `./gradlew rsdroid-testing:assembleRelease` as you indicate
- get Anki-Android clean+up-to-date and alter this dep to point to local rsdroid-testing
- run `./gradlew jacocoUnitTestReport` to make sure things run
- add whatever unit test you need, optionall print debugging in rsdroid ;-), rebuild either/both dep and rerun test

repeat as needed until fault is isolated ?

except currently there is apparently some other alteration to make in RobolectricTest? That's the part I'd like to avoid if possible

**Comment by mikehardy** [comment]
I can imagine all tests pass, it looks really close. Most of my comments were about the last percent (or similar) of removing a couple TODOs, bike-shedding on a variable name or annotation etc

File: AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java (hunk 2 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java]: I'd actually renamed this from `useBackend` just before pushing. Happy to go back.

Agreed on the other comment

File: AnkiDroid/build.gradle (hunk 4 above)
Comment (david-allison-1) [reply AnkiDroid/build.gradle:254]: Yep - `anki-android-backend` defines the Java API/JNI ABI, so this will likely need to be updated if changes should be tested

`anki-android-testing` just contains the `.so` for Robolectric, so if changes are made on the Rust side, a new version will also be required (typically best to change both dependencies at the same time to reduce confusion). 

I've added `RustBackendLoader.loadRsdroid(path)`, which would be a single line replacement in `RobolectricTest.java`

Will get this into backend 0.1.0

**Comment by david-allison-1** [comment]
Force pushed. Ready for re-review

**Review by mikehardy — APPROVED** [review]
Cleaner in a few ways, LGTM - I'll approve but I can't possibly hit the buttons on this one. You should have the honor - merge at will and hit that publish button

What could possibly go wrong

/ducks :rofl: but seriously launch it!
